llvm/tools/llvm-foreach/llvm-foreach.cpp

```diff
@@ -0,0 +1,207 @@
+
+using namespace llvm;
+
+static cl::list<std::string> InputFileLists{
+    "in-file-list", cl::OneOrMore,
+    cl::desc("Input list of file names, file names must be delimited by a "
+             "newline character."),
+    cl::value_desc("filename")};
+
+static cl::list<std::string> InputCommandArgs{
+    cl::Positional, cl::OneOrMore, cl::desc("<command>"),
+    cl::value_desc("command")};
+
+static cl::list<std::string> Replaces{
+    "in-replace", cl::OneOrMore,
+    cl::desc("Specify input path in input command, this will be replaced with "
+             "names read from corresponding input list of files."),
+    cl::value_desc("R")};
+
+static cl::opt<std::string> OutReplace{
+    "out-replace",
+    cl::desc("Specify output path in input command, this will be replaced with "
+             "name of temporary file created for writing command's outputs."),
+    cl::init(""), cl::value_desc("R")};
+
+static cl::opt<std::string> OutDirectory{
+    "out-dir",
+    cl::desc("Specify directory for output files; If unspecified, assume "
+             "system temporary directory."),
+    cl::init(""), cl::value_desc("R")};
+
+static cl::opt<std::string> OutFilesExt{
+    "out-ext",
+    cl::desc("Specify extenstion for output files; If unspecified, assume "
+             ".out"),
+    cl::init("out"), cl::value_desc("R")};
+
+// Emit list of produced files for better integration with other tools.
+static cl::opt<std::string> OutputFileList{
+    "out-file-list", cl::desc("Specify filename for list of outputs."),
+    cl::value_desc("filename"), cl::init("")};
+
+static void error(const Twine &Msg) {
+  errs() << "llvm-foreach: " << Msg << '\n';
+  exit(1);
+}
+
+static void error(std::error_code EC, const Twine &Prefix) {
+  if (EC)
+    error(Prefix + ": " + EC.message());
+}
+
+int main(int argc, char **argv) {
+  cl::ParseCommandLineOptions(
+      argc, argv,
+      "llvm-foreach: Execute specified command as many times as\n"
+      "necessary to use up the list of input items.\n"
+      "Usage:\n"
+      "llvm-foreach --in-file-list=a.list --in-replace='{}' -- echo '{}'\n"
+      "NOTE: commands containig redirects are not supported by llvm-foreach\n"
+      "yet.\n");
+
+  ExitOnError ExitOnErr("llvm-foreach: ");
+
+  if (InputFileLists.size() != Replaces.size())
+    error("Number of input file lists and input path replaces don't match.");
+
+  std::vector<std::unique_ptr<MemoryBuffer>> MBs;
+  std::vector<line_iterator> LineIterators;
+  for (auto &InputFileList : InputFileLists) {
+    std::unique_ptr<MemoryBuffer> MB = ExitOnErr(
+        errorOrToExpected(MemoryBuffer::getFileOrSTDIN(InputFileList)));
+    LineIterators.push_back(line_iterator(*MB));
+    MBs.push_back(std::move(MB));
+  }
+
+  SmallVector<StringRef, 8> Args(InputCommandArgs.begin(),
+                                 InputCommandArgs.end());
+
+  if (Args.empty())
+    error("No command?");
+
+  struct ArgumentReplace {
+    int ArgNum = -1;
+    std::string Prefix;
```

Comment:
It is enough to just store an index where `Replace` string starts, and it's length.
No need to split a string into two allocations.

Reply:
Okay, done.